### PR TITLE
Support CPU input for device `QuantileDMatrix`.

### PR DESCRIPTION
**Pull request by trivialfis**
- Copy `GHistIndexMatrix` to `Ellpack` when needed.

### CODE REVIEW

File: src/data/ellpack_page.cu [hunk 1]
```diff
@@ -284,6 +286,77 @@ EllpackPageImpl::EllpackPageImpl(AdapterBatch batch, float missing, int device,
...
+
+  dh::LaunchN(row_stride * page.Size(), [=] __device__(size_t idx) mutable {
+    auto ridx = idx / row_stride;
+    auto ifeature = idx % row_stride;
+
+    auto r_begin = d_row_ptr[ridx];
+    auto r_end = d_row_ptr[ridx + 1];
+    size_t rsize = r_end - r_begin;
+
+    if (ifeature >= rsize) {
```
Comment (RAMitchell) [inline src/data/ellpack_page.cu:310]: Can't you set the null values here?

File: src/data/iterative_dmatrix.h [hunk 2]
```diff
@@ -22,7 +22,31 @@ class HistogramCuts;
...
+ * matrix `SparsePage`. However, it has its limitation (can be fixed if needed):
+ *
+ * - It's only supported by hist tree method (both CPU and GPU) since approx requires a
+ *   re-calculation of quantiles for each iteration. We can fix this by retaining a
+ *   reference to the callback if there are feature requests.
+ *
+ * - The CPU format and the GPU format are different, the former uses a CSR + CSC for
+ *   histogram index while the latter uses only Ellpack. This results into a design that
+ *   we can obtain the GPU format from CPU but not the other way around since we can't
+ *   recover the CSC from Ellpack. More concretely, if users want to construct a CPU
```
Comment (RAMitchell) [inline src/data/iterative_dmatrix.h:44]: What is the problem with obtaining CSC from ellpack?

Comment (trivialfis) [reply src/data/iterative_dmatrix.h:44]: How to get the feature index for each element from ellpack?

File: src/data/ellpack_page.cu (hunk 1 above)
Comment (trivialfis) [reply src/data/ellpack_page.cu:310]: Thanks for the suggestion. Let me try to merge the kernels.

File: src/data/iterative_dmatrix.h (hunk 2 above)
Comment (trivialfis) [reply src/data/iterative_dmatrix.h:44]: Didn't think of that. Yup, we can recover the feature index by binary searching the cut values. Will update the comment and leave it for a different PR for copying data from GPU to CPU.

File: src/data/ellpack_page.cu (hunk 1 above)
Comment (trivialfis) [reply src/data/ellpack_page.cu]: Merged.

**Review by RAMitchell — APPROVED** [review]

**Pull request merged**